### ISSUE TITLE
Difference in gini values

### ISSUE THREAD

**Issue by taposh** [mmngreco/IneqPy#8]
Hi I am planning to use gini from ineqpy but the value is different when I compare it to developed code. Can you let me know where I am wrong?

```
#https://www.kaggle.com/batzner/gini-coefficient-an-intuitive-explanation
#Note: I have created a pandas data frame with predicted values and actual results
#pred is values predicted and result is actual or Truth

pred, result = final_result.pred.values, final_result.result.values
def gini(actual, pred):
    assert (len(actual) == len(pred))
    all = np.asarray(np.c_[actual, pred, np.arange(len(actual))], dtype=np.float)
    all = all[np.lexsort((all[:, 2], -1 * all[:, 1]))]
    totalLosses = all[:, 0].sum()
    giniSum = all[:, 0].cumsum().sum() / totalLosses
    giniSum -= (len(actual) + 1) / 2.
    return giniSum / len(actual)

def gini_normalized(actual, pred):
    return gini(actual, pred) / gini(actual, actual)

gini_predictions = gini(result, pred)
gini_max = gini(result, result)
ngini= gini_normalized(result, pred)
print('Gini: %.3f, Max. Gini: %.3f, Normalized Gini: %.3f' % (gini_predictions, gini_max, ngini))

#using ineqpy
import ineqpy as inq
measure = inq.api.Survey(final_result, weights='pred')
print('Gini using ineqpy: %.3f' % measure.gini('result'))
```

**Kaggle's method** Gini: 0.131, Max. Gini: 0.382, Normalized Gini: 0.344
**Ineqpy**  Gini using ineqpy: 0.504

**Comment by mmngreco** [comment]
Hello,
I'm sorry, I haven't had much time to review it and I've never seen that use case and I find it interesting. What I have seen is the following:

If I understood correctly, they are different algorithms. The one you mention seems to compare two vectors, while the one I have implemented follows the original [definition](https://en.wikipedia.org/wiki/Gini_coefficient) for a single vector. The weights variable refers to the frequency (absolute or relative) of each element of the vector. 

A weight vector that includes zeros is like telling it not to take those observations into account.

You would have to adapt the inputs so that you could quantify that. I would have to think a little about how you could adapt the inputs to do that.

If you solve it, I'll be happy to see how you've done it.

If I can ask, what is the use of the gini coefficient that you want to do?

**Comment by taposh** [comment]
As you might have read in the link from Kaggle (https://www.kaggle.com/batzner/gini-coefficient-an-intuitive-explanation), we are using gini as a metric to measure our prediction.

At a high level :
Instead of going through the population from poorest to richest, we go through our predictions from lowest to highest. Instead of summing up the income, we sum up the actual values of our predictions, which leads to the development of Lorenz Curve. We normalize both axes so that they go from 0 to 100% like in the economic figure and display the 45° line for illustrating random guessing. Finally the area between the 45° line and Lorenz Curve is used as gini to compare results.

Let me know what are your thoughts. I would love to set-up a meeting if possible on phone/webex to get your inputs.

**Comment by taposh** [comment]
@mmngreco would you like to comment here..

**Comment by mmngreco** [comment]
Hi @taposh, that's the classic definition I've always used but I've never applied it to a classification problem like the one you propose. Your case uses binary data, which is why the algorithm implemented in ineqpy [see 2] is not what you should use. 

I've been reading and arguing with a colleague and it seems that gini could be seen as the special case of order 1 cross-entropy in classification problems. 

An interesting relationship is [see 1]: 

Gini = 2 × AUC - 1

Which makes us think that maybe logloss and gini (in classification problems) are approximations of the same thing.

To conclude: the ineqpy algorithm is not intended for classification problems. The algorithm you have seems to do what it should. An alternative could be to use the logloss function, or the relationship you have above. 

I hope it helps, below I leave you some interesting references that I have consulted.

### References
1. https://en.wikipedia.org/wiki/Receiver_operating_characteristic#Area_under_the_curve
1. https://www.ntanet.org/NTJ/41/1/ntj-v41n01p139-42-calculating-income-distribution-indices.pdf?v=%CE%B1&r=19426893815398216
1. https://ml-cheatsheet.readthedocs.io/en/latest/loss_functions.html
1. https://en.wikipedia.org/wiki/Gini_coefficient#Other_approaches
1. https://en.wikipedia.org/wiki/Cross_entropy#Cross-entropy_error_function_and_logistic_regression
